doc/man3/OSSL_PARAM_construct_from_text.pod

```diff
@@ -0,0 +1,167 @@
+value size I<value_n>, and given a parameter descriptor array
+I<paramdefs>, it converts the value to something suitable for
+L<OSSL_PARAM(3)> and stores that in the buffer I<buf>, and modifies
+the parameter I<to> to match.
+I<buf_n>, if not NULL, will be assigned the number of bytes used in
+I<buf>.
+If I<buf> is NULL, only I<buf_n> will be modified, everything else is
+left untouched, allowing a caller to find out how large the buffer
+should be.
+I<buf> needs to be safely aligned for all kinds of uses, including as
+native type data.
+
+OSSL_PARAM_allocate_from_text() works like OSSL_PARAM_construct_from_text(),
+except it allocates the buffer internally.
+The caller must remember to free the data of I<to> when it's not
+useful any more.
+
+For parameters having the type B<OSSL_PARAM_INTEGER>,
+B<OSSL_PARAM_UNSIGNED_INTEGER>, or B<OSSL_PARAM_OCTET_STRING>, both
+functions will interpret the I<value> differently if the key starts
+with "hex".
+In that case, the value is decoded first, and the result will be used
+as parameter value.
+
+=head1 RETURN VALUES
+
+OSSL_PARAM_construct_from_text() and OSSL_PARAM_allocate_from_text()
+returns 1 on success, and 0 on error.
+
+=head1 NOTES
+
+The parameter descriptor array comes from functions dedicated to
+return them.
+The following B<OSSL_PARAM> attributes are used:
+
+=over 4
+
+=item I<key>
+
+=item I<data>
+
+=item I<data_size>
+
+=back
+
+All other attributes are ignored.
+
+The I<data_size> attribute can be zero, meaning that the parameter it
+describes expects arbitrary length data.
+
+=head1 EXAMPLE
+
+Code that looked like this:
+
+  int mac_ctrl_string(EVP_PKEY_CTX *ctx, const char *value)
+  {
+      int rv;
+      char *stmp, *vtmp = NULL;
+      stmp = OPENSSL_strdup(value);
+      if (!stmp)
+          return -1;
+      vtmp = strchr(stmp, ':');
+      if (vtmp) {
+          *vtmp = 0;
+          vtmp++;
+      }
+      rv = EVP_MAC_ctrl_str(ctx, stmp, vtmp);
+      OPENSSL_free(stmp);
+      return rv;
+  }
+
+  ...
+
+
+  char *macopt;
+  for (i = 0; i < sk_OPENSSL_STRING_num(macopts); i++) {
+      macopt = sk_OPENSSL_STRING_value(macopts, i);
+      if (pkey_ctrl_string(mac_ctx, macopt) <= 0) {
+          BIO_printf(bio_err,
+                     "MAC parameter error \"%s\"\n", macopt);
+          ERR_print_errors(bio_err);
+          goto mac_end;
+      }
+  }
+
+Can be written like this instead:
+
+  OSSL_PARAM *params =
+      OPENSSL_zalloc(sizeof(OSSL_PARAM)
+                     * (sk_OPENSSL_STRING_num(opts) + 1));
+  const OSSL_PARAM *paramdefs = EVP_MAC_CTX_set_param_types(mac);
+  size_t params_n;
+
+  for (params_n = 0; params_n < (size_t)sk_OPENSSL_STRING_num(opts);
+       params_n++) {
+      char *opt = sk_OPENSSL_STRING_value(opts, (int)params_n);
+      char *stmp, *vtmp = NULL;
+
+      if ((stmp = OPENSSL_strdup(opt)) == NULL
+          || (vtmp = strchr(stmp, ':')) == NULL
+          || (*vtmp++ = 0) /* Always zero */
```

Comment:
Hmmmm.....

I'm not keen on this trick in example code.

Reply:
It makes for pretty straight forward code, though...  the alternative would probably be a bit convoluted, and risks taking away more attention from what the example is really here to demonstrate.

Reply:
I'll accept that a little grudgingly.

Shouldn't it be '\0' not 0 :)

Reply:
> Shouldn't it be '\0' not 0 :)

If you really want to be utterly strict about it, yes.  In reality, it matters very little.  However, if that makes you a little less grudgy (i.e. compensate for that other thing), I'll happily make that change.

Reply:
(and no, I'm not sure "grudgy" is an actual word...  or maybe I just coined it ;-))

Reply:
That's horrible.  You can make the code clean and fitting into the OpenSSL style by using a goto err.  You really should fix it.

Reply:
#9577